### ISSUE TITLE
sftp.realPath: No such file Error when there is definitely a file.

### ISSUE THREAD

**Issue by johnmeister** [theophilusx/ssh2-sftp-client#179]
Node --version 12.10.0
typscript -- version 3.6.2
ssh2-sftp-client 4.2.1

Error: sftp.realPath: No such file /Outgoing/CommissionStmts/00001.pdf

The above error is thrown every time I attempt to get/ fastget/stat/exists any files.
I have visually inspected that the file & file paths are correct and that I have the correct permissions I did this by using filezilla to log into the server and pull the file from the same IP address using the same set of Credentials.
below is a extract from my code:

-index.ts
`import Client from "ssh2-sftp-client";`

`class CommissionStatement{`
`    ftp:Client = new Client();`
`    public async GetFile(){`
`        await this.ftp.connect({`
`            host: "sftp.host.com",`
`            port: 22,`
`            username: "user",`
`            password: "pass",`
`            readyTimeout: 65000`
`         });`

`        let remoteDir = '/Outgoing/CommissionStmts/';`

`        let list = await this.ftp.list(remoteDir );`
`        /* this is to prove I can access the directory, it's also known that I have all permissions for all files and directories. */`
`        console.log(list[0]);`

`        let remoteFullPath = remoteDir+"00001.pdf"`
`        let file = await this.ftp.get(remoteFullPath ,localFilePath).catch((error)=>{throw error})`
`        /*the above line throws the error everytime.*/`

`    }// end GetFile()`
`}//end class`

**Comment by theophilusx** [comment]
This is very odd, especially if the file is showing up in the result
from list(). If you do a

console.dir(list);

instead of

console.log(list[0]);

what does it show?

Also, what happens if you change he get() to be

let file = await this.ftp.fastGet(remoteFullPath ,localFilePath);

and finally, can you show the actually JS rather than the typescript?

John Meister <notifications@github.com> writes:

> Node --version 12.10.0
> typscript -- version 3.6.2
> ssh2-sftp-client 4.2.1
>
> Error: sftp.realPath: No such file /Outgoing/CommissionStmts/00001.pdf
>
> The above error is thrown every time I attempt to get/ fastget/stat/exists any files.
> I have visually inspected that the file & file paths are correct and that I have the correct permissions I did this by using filezilla to log into the server and pull the file from the same IP address using the same set of Credentials.
> below is a extract from my code:
>
> -index.ts
> `import Client from "ssh2-sftp-client";
>
> class CommissionStatement{
> ftp:Client = new Client();
> public async GetFile(){
> await this.ftp.connect({
>             host: "sftp.host.com",
>             port: 22,
>             username: "user",
>             password: "pass",
>             readyTimeout: 65000
> });
> let remoteDir = '/Outgoing/CommissionStmts/';
> let list = await this.ftp.list(remoteDir );
> /* this is to prove I can access the directory, it's also known that I have all permissions for all files and directories. */
> console.log(list[0]);
> let remoteFullPath = remoteDir+"00001.pdf"
> let file = await this.ftp.get(remoteFullPath ,localFilePath).catch((error)=>{throw error})
> /*the above line throws the error everytime.*/
>
> }
> }`


-- 
Tim Cross

**Comment by johnmeister** [comment]
`console.dir(list)`
 results: 
`[
  {
    type: '-',
    name: '00001.pdf',
    size: 71119,
    modifyTime: 1551414758000,
    accessTime: 1551414758000,
    rights: { user: 'rw', group: 'rw', other: 'rw ' },
    owner: 0,
    group: 0
  }]`
any method that try's to access the file directly throws the same error.
tried
`console.log(await ftp.stat(file))` 
throws Error: sftp.realPath: No such file /Outgoing/CommissionStmts/00001.pdf
      at formatError (node_modules\ssh2-sftp-client\src\index.js:60:10)
tried
`await this.ftp.exists('/Outgoing/CommissionStmts/00001.pdf')` returns false 
** for sanity I ran:
`await this.ftp.exists('/Outgoing/CommissionStmts/)` which correctly returns "-d"
I cant give you the compiled code in js. business rules wont allow it.
`await this.ftp.fastGet('/Outgoing/CommissionStmts/00001.pdf',join(__dirname,'00001.pdf'))`
throws  Error: sftp.realPath: No such file /Outgoing/CommissionStmts/00001.pdf
      at formatError (node_modules\ssh2-sftp-client\src\index.js:60:10)

**Comment by johnmeister** [comment]
Just tried using the same code and changing the connection strings to a different FTP site and it was transferring files as expected. Could this be a permission issue?

**Comment by theophilusx** [comment]
The only odd thing I notice in that output is the user and group ID's
are both 0, which would indicate the files are owned by root and in the
group root or you are on a windows file system (for the sftp server),
What do you get if you do a

let absPath = await this.ftp.realPath('.');
console.log(absPath);

?
John Meister <notifications@github.com> writes:

> `console.dir(list)`
>  results: 
> `[
>   {
>     type: '-',
>     name: '00001.pdf',
>     size: 71119,
>     modifyTime: 1551414758000,
>     accessTime: 1551414758000,
>     rights: { user: 'rw', group: 'rw', other: 'rw ' },
>     owner: 0,
>     group: 0
>   }]`
> any method that try's to access the file directly throws the same error.
> tried
> `console.log(await ftp.stat(file))` 
> throws Error: sftp.realPath: No such file /Outgoing/CommissionStmts/00001.pdf
>       at formatError (node_modules\ssh2-sftp-client\src\index.js:60:10)
> tried
> `await this.ftp.exists('/Outgoing/CommissionStmts/00001.pdf')` returns false 
> ** for sanity I ran:
> `await this.ftp.exists('/Outgoing/CommissionStmts/)` which correctly returns "-d"
> I cant give you the compiled code in js. business rules wont allow it.
> `await this.ftp.fastGet('/Outgoing/CommissionStmts/00001.pdf',join(__dirname,'00001.pdf'))`
> throws  Error: sftp.realPath: No such file /Outgoing/CommissionStmts/00001.pdf
>       at formatError (node_modules\ssh2-sftp-client\src\index.js:60:10)


-- 
Tim Cross

**Comment by johnmeister** [comment]
`let absPath = await this.ftp.realPath('.');
console.log(absPath);`

returns "/"

**Comment by theophilusx** [comment]
Yes, my guess is that it is a permissions issue or some sort of network issue with the remote server. The listing for the file you are trying to retrieve indicates it is owned by root and in the group root, which wold be very unusual. You certainly wouldn't/shouldn't be connecting as root due to the huge security issues that would present. Whatever the cause, I think this is some server issue rather than an issue with the code or ssh2-sftp-client. Things I would try

1. Use just command line sftp and see how you go. Filezilla has various options/smarts built in which may be dealing with the situation which a basic command line sftp client, like openssh sftp might reveal.  You can also use the exact same authentication (either password or key file).  In particular, filezilla can hide some network issues which more basic sftp clients will reveal. In particular, lost connections or partial downloads. 

2. I would get a list() for all the directories in the path i.e. start with list('/'), then list('/Outgoing') etc and see what the permissions are on each directory  

3. Try creating a directory on the remote server at different levels in the hierarchy with ftp.mkdir() and compare the group and user at each level.

**Comment by theophilusx** [comment]
BTW, do you know what the remote sftp software is? There have been reports of problems with BRICKftp. There are some poor sftp servers out there which filezilla is able to work-around, but other libraries, like ssh2 and sftp-streams cannot.

**Comment by johnmeister** [comment]
I dont. I can try to find out but it's a third party so who knows how forthcoming they will be with that type of info.
Ill write some quick tests tonight/ tomorrow based on your troubleshooting tips above.

**Comment by johnmeister** [comment]
` async testMethod() {
        await this.ftp.connect({
            host: "sftp.host.com",
            port: 22,
            username: "user",
            password: "pass",
            readyTimeout: 65000,
        });
        let p = [this.ftp.mkdir("/PLICATest", false), this.ftp.mkdir("/Outgoing/PLICATEST", false)]
        let b: string[] = await Promise.all(p).catch((err) => {
            console.log(err)
            return [err, err]
        })
        console.log(b[0], b[1]);
        await this.ftp.end()
        return
    }`

**this prints `/PLICATest directory created /Outgoing/PLICATEST directory created` as expected.
and When I compare its file permissions they are exactly the same as the other directories.
Ive also tried putting files and pulling them from here with no luck.**
 
**PLICATest**
 `{
    type: 'd',
    name: 'PLICATest',
    size: 0,
    modifyTime: 1571863150000,
    accessTime: 1571862565000,
    rights: { user: 'rw', group: 'rw', other: 'rw ' },
    owner: 0,
    group: 0
  },`
 **compared to /Outgoing**
`  {
    type: 'd',
    name: 'Outgoing',
    size: 0,
    modifyTime: 1571863133000,
    accessTime: 1527010264000,
    rights: { user: 'rw', group: 'rw', other: 'rw ' },
    owner: 0,
    group: 0
  },`
**and the same for /Incoming**
`{
    type: 'd',
    name: 'Incoming',
    size: 0,
    modifyTime: 1530221370000,
    accessTime: 1527013841000,
    rights: { user: 'rw', group: 'rw', other: 'rw ' },
    owner: 0,
    group: 0
  },`

**then I did this:**
`        let p = await this.ftp.list('.')
        let b = await this.ftp.list('/Outgoing/CommissionStmts')
        let e = await this.ftp.exists('/Outgoing/CommissionStmts/00001pdf')

        console.dir(p)
        console.dir(b)
        console.dir(e)`

**results are**
` [...], [  {
    type: '-',
    name: '00001.pdf',    
    rights: { user: 'rw', group: 'rw', other: 'rw ' },
    owner: 0,
    group: 0
  }], false`

**Comment by theophilusx** [comment]
I have added a very simple script called /example/test-get-writeable.js to the repository. You would call it as

node ./test-get-writeable.js /Outgoing/CommissionStmts/00001.pdf

If it works, you know the issue is with your code. If it fails, then the issue is with the sftp server. I have tested this against some local and remote test servers using Node 12.11.0 with no errors. See inside the script for some notes and a config setting for the host, username and password.

**Comment by johnmeister** [comment]
I downgraded my library version to 2.4.3 and now it works without issue.

**Comment by theophilusx** [comment]
Which library did you download to 2.4.3?

**Comment by theophilusx** [comment]
It is tricky because I canot reproduce this issue on any server I've
tried. Looking at the realpath() code, it seems that the 'No such file'
error will only get thrown when a component in the path is missing
i.e. in /a/b/c, an exception only gets thrown when '/', 'a' or 'b' is
unreadable/missing. If 'c' is missing, no exception is raised.

I will try writing a simple script which only uses the ssh2/ssh2-streams
libraries. If we can demonstrate the same behaviour and sow a difference
between versions, we at least know the issue is in one of the upstream
libraries.

John Meister <notifications@github.com> writes:

> I'm happy to help find it and resolve it. just let me know what I can do to help.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or unsubscribe.


-- 
Tim Cross

**Comment by theophilusx** [comment]
I have created two JS scripts which might help identify any differences. In the tools folder of the repository, there is a script called lowlevel-realpath.js, which just tests the result of calling realpath directly in the ssh2 module, bypassing ssh2-sftp-client module. Call it like

`node lowlevel-realpath.js /path/to/test`

In the examples directory, there is also a script called realpath.js, which does the same thing, but via calling the realPath() function is ssh2-sftp-client.  Call it in the same way. 

node realpath.js /pathto/test

Both of these scripts should give the same results. It would be good to test them against the server that was giving the error before. If the first one also fails, we know the issue is with the ssh2 module. If only the second one gives the error, then the issue is with ssh2-sftp-client. It may also be useful to try testing with arguments of '.' and '..' and see what results we get. The first should report the current remote working directory and the second should report the parent of the remote working directory. 

It would be good to test with current version and the version 2.4.1 which seems to be working. 

Note that you may need to modify the require statements for the modules, depending on how your setup is configured. Also note that I use the dotenv module and use .env file to store the username/password and host details. You may want to hard code these ifyou don't want to use dotenv.

**Comment by theophilusx** [comment]
A bug has been found in the exist() method which *might* impact on this issue. Would be great to know if the same issue exists with v4.2.3. Any chance you could try out that version?

**Comment by theophilusx** [comment]
As there has been no response and it is likely the bug already identified and resolved has also addressed this issue, I am closing this issue.

**Comment by EduardoIbarra** [comment]
Hello, the very same bug is happening to me. 

> ssh2-sftp-client-4.3.1

**Comment by theophilusx** [comment]
Can you provide -

1. What platform are you running on (Linux, Mac, Windows?)
2. Minimal code example which reproduces the issue
3. Try running the tools/lowlevel-realpath.js and example/realPath.js scripts

Note you will either need to use the dotenv package and setup a .env file or modify the two scripts in order to run them.

**Comment by theophilusx** [comment]
Also if you wanted to, you could try using the release-5.0.0 branch. This will be the next released version and has some error handling enhancements which *might* affect the issue you are seeing.

**Comment by jpbecotte** [comment]
I had a similar issue (instead of File not found, it was «Failure»). I could list the files, but any attempt to stat or get any files in the list would result in this error. Like @johnmeister , I reverted to 2.4.3 and it is working perfectly.

**Comment by theophilusx** [comment]
> I had a similar issue (instead of File not found, it was «Failure»). I
> could list the files, but any attempt to stat or get any files in the
> list would result in this error. Like @johnmeister , I reverted to
> 2.4.3 and it is working perfectly.

Errors which just say "Failure" are errors originally raised inside the
SSH2/SSH2Streams modules used by ssh2-sftp-client. This would indicate
the issue is with those underlying modules. There is a script in the
tools sub-directory of the git repository called lowlevel-realpath.js,
which performs a realpath check using only the underlying
SSH2/SSH2Streams modules. If you get an error using this script, it
confirms the issue is with those modules and needs to be logged
upstream. Look in package.json to see what underlying versions of SSH2
is being used for current and 2.4.3 releases. 

Note that there are some known significant bugs in the 2.4.3 version. If
it works for you great, but there are risks due to these known issues.

Version 5.0.0 has seen significant improvements in error checking and
error reporting. It would be worth testing it and see if you still see
the same issue. It is available in the release-5.0.0 branch of the git
repo.

**Comment by theophilusx** [comment]
I cannot reproduce this issue. Can you provide a minimal example that demonstrates the issue? Also, if you feel so inclined, I'm currently testing release 5.0.0, which has some significant internal changes relating to error checking and error reporting which might either avoid the issue or provide more meaningful error information than just 'failure'. It is in branch release-5.0.0 of the git repo

**Comment by EduardoIbarra** [comment]
Hello, I've just installed 5.0.0 release branch by:
`$ npm install git://github.com/theophilusx/ssh2-sftp-client.git#release-5.0.0 --save`
But I am getting the same error:
```
(node:1921) UnhandledPromiseRejectionWarning: Error: fastGet: No such file: /ELY/WIPsummary/cmif_SLF_ELYRJ5.200105.xml
    at Object.formatError (/cielopuerto/node_modules/ssh2-sftp-client/src/utils.js:62:18)
    at SftpClient.fastGet (/cielopuerto/node_modules/ssh2-sftp-client/src/index.js:509:21)
    at process._tickCallback (internal/process/next_tick.js:68:7)
(node:1921) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 287)
```

This is my code:

```
const sftp = new Client();
const tmpFolder = 'tmp';
const path = '/ELY/WIPsummary';

const listSFTPFiles = async () => sftp.connect({
  host: 'sftp.theserver.com',
  port: '22',
  username: 'myuser',
  password: '*******,
}).then(() => sftp.list(path)).then((data) => {
  data.forEach((remoteFile) => {
    // console.log(remoteFile); <--- This prints correctly each file on the directory
    sftp.fastGet(`${path}/${remoteFile.name}`, `./${tmpFolder}/${path}/${remoteFile.name}`);
  });
}).catch((err) => {
  console.log(err, 'catch error');
});
listSFTPFiles();
```

SFTP server is hosted on a Windows machine (I do not have more details on this).

**Comment by theophilusx** [comment]
Part of your problem is the foreach loop. You are firing off multiple
fastget() requests before each has finished. This is also why your error
is being reported as an uncaught exception - there is nothing there to
catch any errors. (as your not 'returning' a promise in that then()
block, it is not part of the chain and so the .catch() at the end won't
help)

For situations where you want to get multiple files like this, you
really want to use async/await and have an await for each fastGet()
call. have a look at some of the examples in the example directory of
the repository. Either wrap it in a try/catch or wrap each call to
fastGet in a try/catch (if you want to catch errors for individual
failures).

If you don't want to use async/await, then you need to use something
like Promise.all() e.g.

...
then(data => {
  let p = []
  data.forEach(f => {
    p.push(sftp.fastGet(`${path}/f.name`, `${tmpdir}/$[path}/f.name`));
  });
  return Promise.all(p);
})
.then(rslts => {
  // rslts will be an array of results - one for each promise
  (...);
  return sftp.end();
})
.catch(err => {
  console.error(...);
});

If it still gives errors about file not being found, it could be a path
separator issue i.e. remote path might need to be ${path}\f.name instead
of ${path}/f.name. This *should* not be necessary, but worth a try just
in case (I don't use Windows and have not worked on that platform since
MS-DOS 3.11!).

**Comment by EduardoIbarra** [comment]
Thanks for your input @theophilusx , I simplified and added async/await try/catch as you suggested with the same result.

```
const Client = require('ssh2-sftp-client');
const sftp = new Client();

const listSFTPFiles = async () => sftp.connect({
  host: 'sftp.theserver.com',
  port: '22',
  username: 'myuser',
  password: '*******,
}).then(async () => {
  try {
    // did not work with forwardslashes either
    await sftp.fastGet('ELY\\WIPsummary\\cmif_SLF_ELAN.200108.xml', './tmp/cmif_SLF_ELAN.200108.xml', null); 
  } catch (err) {
    console.log(err);
  }
});
listSFTPFiles();
```
The error:

> { Error: fastGet: No such file: ELY\WIPsummary\cmif_SLF_ELAN.200108.xml
>     at Object.formatError (/home/ec2-user/Projects/cielopuerto/node_modules/ssh2-sftp-client/src/utils.js:62:18)
>     at SftpClient.fastGet (/home/ec2-user/Projects/cielopuerto/node_modules/ssh2-sftp-client/src/index.js:509:21)
>     at process._tickCallback (internal/process/next_tick.js:68:7) code: 'ENOENT', custom: true }

I made sure the file exists and that I am using the same format/path as the list() method gives me. Also I connected using a gui client to list and download files with the same connection parameters as in my code. So I think this might be a Windows incompatibility issue. Thank you for all the help!

**Comment by theophilusx** [comment]
I really don't know if you should mix up async/await inside promise
chains - really should be one or the other.

I have reports of other people using the module with a windows server
OK. You might get some success playing around with some of the scripts
in the example directory. I would also try

const Client = require('ssh2-sftp-client');
const sftp = new Client();

const listSFTPFiles = async (src, dst) => {
  try {
    await sftp.connect({
      host: 'sftp.theserver.com',
      port: '22',
      username: 'myuser',
      password: '*******,
    });
    await sftp.fastGet(src, dst); 
  } catch (err) {
    console.log(err);  
  }
};

listSFTPFiles(
  'ELY\\WIPsummary\\cmif_SLF_ELAN.200108.xml',
  './tmp/cmif_SLF_ELAN.200108.xml'
).catch(err => {
  console.log(err);
});

I would also use CLI sftp tool not GUI to confirm things - too many GUI
tools hide details, plus it is usually much easier to get additional
debugging information as to what the client is doing. 

Finally, once the connection is established, check the values of
sftp.remotePathSep and sftp.remotPlatform.

**Comment by theophilusx** [comment]
It is possible this issue was being cased by permission settings on the parent directory. The exists() method has been changed to use stat() and not use realPath(), so this issue should now be resolved. Closing this issue for now as a new version has been released (v5.0.1)). Can re-open it if the issue persists.

**Comment by johndpotts** [comment]
so for what it's worth I just ran into this today. Worked fine with 6.x on a local windows machine, but when running in a lambda it would list the files but error out on fastGet with "file not found". stat and exists both worked though.
downgrading to 2.4.3 fixed the issue and it worked fine in the lambda (and still worked locally), but I will poke around and see if there's a way to resolve that doesn't involve downgrading to a much older version

**Comment by theophilusx** [comment]
I'm assuming you are seeing this with version 7.2.1. Try the latest code
from the repository. It has a couple of bug fixes which while not
directly related to your issue, may provide more info or might just work
better.

There are also some validation scripts in the validate directory of the
repository. These scripts only use ssh2 and don't use any of
ssh2-sftp-client. They can be useful in identifying whether the issue is
due to something in ssh2 or in the ssh2-sftp-client module. You should
be able to use one of the scripts in that directory as a template to
test and determine at what layer the issue is arising.

It is possibly an issue at the ssh2-sftp-client level as v7.x removed a
lot of the checking and validation code from the previous version. This
code was removed to improve overall performance. It was originally added
because in many cases, ssh2 or the target server often provided very
uninformative error messages which were less than helpful to developers.
I added additional validation checks to provide more informative errors,
but unfortunately, that had a performance hit on the more general case
when no errors occurred (because it involved additional queries to the
remote sftp server). I therefore decided to remove some of the
additional validation code.

Version 7.x is also using a completely new re-write of ssh2, so it could
be an ssh2 issue. There were major problems with ssh2 and more recent
node versions which could not be fixed without significant
re-factgoring. The ssh2 author decided it was better to do a complete
re-design and re-write of the code as that would also allow addressing
some other design limitations in the original implementation. It is
possible new bugs have been introduced as part of that
re-implementation. In fact, the reason I've not yet pushed out the most
recent changes to ssh2-sftp-client is because I know some bugs have been
addressed in the most recent versions of ssh2, but a new version has not
yet been pushed and I was waiting on that before pushing out a new
ssh2-sftp-client.
